"No such grain for public ID" after grain stops

I installed Sandstorm on a fresh Ubuntu 14.04 VPS.
1. Create a new grain (tried Etherpad, Wave)
2. Click the Sandstorm logo to return to the home page
3. Wait a few minutes until the grain is killed
4. Click to open the grain again

_Expected result:_ grain is started again

_Actual result:_ message like `No such grain for public ID: yb8d9nh5hfxdvq1slqa7`

The "Restart App" button just generates a log message like

`../src/node-capnp/capnp.cc:1702: error: e = src/kj/async-io.c++:339: error from OS: connect(): Connection refused; toString() = unix:/var/sandstorm/grains/kkDH3cQAZcj67X9rMLesw2/socket`

But refreshing the entire page does restart the grain, and load the correct new hostname. Is the old hostname being cached somewhere in the frontend longer than it should?

--- a/shell/shared/grain.js
+++ b/shell/shared/grain.js
@@ -421,7 +421,7 @@ Router.map(function () {
       var session = Session.get("session-" + grainId);
       if (session === "pending") {
         return result;
-      } else if (session) {
+      } else if (session && Sessions.findOne(session.sessionId)) {
         result.appOrigin = document.location.protocol + "//" + makeWildcardHost(session.hostId);
         setCurrentSessionId(session.sessionId, result.appOrigin, grainId);
         result.sessionId = session.sessionId;

--- a/shell/shared/shell.js
+++ b/shell/shared/shell.js
@@ -59,6 +59,14 @@ if (Meteor.isServer) {
     }
   });
 
+  Meteor.publish("sessions", function () {
+    if (this.userId) {
+      return Sessions.find({userId: this.userId});
+    } else {
+      return [];
+    }
+  });
+
   Meteor.publish("devApps", function () {
     return DevApps.find();
   });
@@ -133,6 +141,7 @@ if (Meteor.isClient) {
 
   Deps.autorun(function () {
     Meteor.subscribe("credentials");
+    Meteor.subscribe("sessions");
   });
 
   makeDateString = function (date) {

